pex/venv/pex.py

```diff
@@ -91,6 +91,40 @@ def _copytree(
             return
 
 
+def _create_shebang(
+    venv_python, # type: str
+):
+    # type: (...) -> str
+    """Return the shebang to be used in the venv pex invoker script.
+
+    When the shebang is longer than the common linux length limit of 128 Bytes we try
+    to replace the shebang with a snippet of shell script which will invoke the correct
+    python interpreter from a path relative to the script itself, which avoids the need
+    to encode the long interpreter as a shebang.
+    """
+
+    # This multiline shebang locates the interpreter in the venv relative to the file this shebang is placed in.
+    # For example the script is usually ~/.pex/venvs/<sha>/<sha>/pex and the venv bin dir is ~/.pex/venvs/<sha>/<sha>/bin/.
+    # The exec command re-execs the venv pex script with the venv interpreter. The unfortunate
+    # trick here is that the shell exec command and its arguemetns are wrapped in quotes so the python
+    # interpreter treats it as a static string when executing the script, essentially ignoring it.
+    long_shebang_exec = '#!{sh_path}\n"exec" "`dirname $0`/bin/{interpreter_basename}" "-sE" "$0" "$@"'
+
+    shebang = "#!{} -sE".format(venv_python)
+    if len(shebang) > 128:
+        sh_path = None
+        # shutil.which only exists in CPython >= 3.3
+        if hasattr(shutil, "which"):
+            sh_path = shutil.which("sh")
```

Comment:
This isn't generally valid. You may be building the PEX on a machine with a different answer than the (or one of the) deployment targets. That said, `/bin/sh` is very, very universally available.

Reply:
My understanding is that this piece of code would always execute on the runtime host during the venv extraction, so the sh_path would be valid on that particular machine. Maybe I'm misunderstand your comment though.

Reply:
Ah, no, you're right. I got confused. You can disregard my wrong machine comments. That said, the fact that `--sh-boot` has lower latency anyhow will likely trump unless you find some issue with using that in your environment.

Reply:
For the particular situation where I ran into this we really want to execute from a pex built venv not an unzipped pex so I don't think --sh-boot will be enough, though I do know we have folks at twtr that want to use --sh-boot in different contexts :)

Reply:
--sh-boot works for both --venv and ZIPAPP PEXes. I'm confused. Are you manually pre-building venvs with `PEX_TOOLS=1 my.pex venv create/venv/here`? If so you can just `create/venv/here/bin/python create/venv/here/pex`

Reply:
You're right that --sh-boot does work with venv. sorry for the confusion. However it looks like it relies on the same [venv script shebang](https://github.com/pantsbuild/pex/blob/c15508ad1a2d48bdef9fbac677dbfa32927e6625/pex/bin/sh_boot.py#L182-L187), so I think in its current form this problem will still exist with --sh-boot. 

We aren't manually creating the venvs, we're relying on the venvs to bootstrap themselves so they can be distributed as a single executable.

Reply:
Ok. Let me finish my test of modifying `--sh-boot`. If that works, it will be more robust and faster.

Reply:
Alright - it looks like that works just fine. I'll post a PR since that was a bug anyhow since the shebang length limit was avoided for the other cases in that script, just not venv. I'll include a test that sets up a 10k venv path or some such.

Reply:
Ok, #1828. I have a robust test there for Linux. Seeing if my technique, which relies on a few errno symbolic constants being the same errnos for shebang too long and file name too long, works on Mac. It does work on Linux.

Reply:
Alright, took a while to get the generic shebang too long testing code working on Mac and Linux, but that's sorted. I'll polish up #1828 for review and I think that should satisfy your concerns++. It actually directly fixes the non --sh-boot case as well, but you'll probably want to switch to --sh-boot if your deployment machines all have `/bin/sh` since it will shave ~50ms of startup overhead and give you native venv speed + ~1ms.